File: apps/macos/Sources/Library/VideoMetadataReader.swift
import AVFoundation
import Foundation

nonisolated struct VideoMetadata: Sendable {
    let duration: Double
    /// Display size after the preferred track transform, in pixels.
    let pixelSize: CGSize?
}

/// Validates a candidate video and extracts the metadata the library needs.
///
/// Validation is deliberately strict: a file that reaches a `WallpaperSession`
/// must be playable, so the failure is reported at import time rather than as a
/// black desktop later.
nonisolated enum VideoMetadataReader {
    static func read(url: URL, fileManager: FileManager = .default) async throws -> VideoMetadata {
        let name = url.lastPathComponent

        guard fileManager.fileExists(atPath: url.path) else {
            throw WallpaperError.fileMissing(name: name)
        }

        let asset = AVURLAsset(
            url: url,
            options: [AVURLAssetPreferPreciseDurationAndTimingKey: true]
        )

        let isPlayable: Bool
        let hasProtectedContent: Bool
        do {
            (isPlayable, hasProtectedContent) = try await asset.load(.isPlayable, .hasProtectedContent)
        } catch {
            throw WallpaperError.notPlayable(name: name)
        }

        guard !hasProtectedContent else { throw WallpaperError.protectedContent(name: name) }
        guard isPlayable else { throw WallpaperError.notPlayable(name: name) }

        let videoTracks = (try? await asset.loadTracks(withMediaType: .video)) ?? []
        guard let track = videoTracks.first else {
            throw WallpaperError.noVideoTrack(name: name)
        }

        let duration = try await asset.load(.duration).seconds
        guard duration.isFinite, duration > 0 else {
            throw WallpaperError.emptyDuration(name: name)
        }

        let pixelSize = try? await Self.displaySize(of: track)
        return VideoMetadata(duration: duration, pixelSize: pixelSize)
    }

    /// Natural size corrected for rotation metadata, so portrait videos report
    /// portrait dimensions.
    private static func displaySize(of track: AVAssetTrack) async throws -> CGSize {
        let (naturalSize, transform) = try await track.load(.naturalSize, .preferredTransform)
        let transformed = naturalSize.applying(transform)
        return CGSize(width: abs(transformed.width), height: abs(transformed.height))
    }
}

File: apps/windows/src/Wallpache.App/Library/VideoMetadataReader.cs
using System;
using System.IO;
using System.Threading.Tasks;
using Wallpache.App.Support;
using Windows.Media.Editing;
using Windows.Storage;

namespace Wallpache.App.Library;

public readonly record struct VideoMetadata(double Duration, int? Width, int? Height);

/// <summary>
/// Validates a candidate video and extracts the metadata the library needs.
///
/// Validation is deliberately strict: a file that reaches a
/// <c>WallpaperSession</c> must be playable, so the failure is reported at
/// import time rather than as a black desktop later.
/// </summary>
public static class VideoMetadataReader
{
    /// <summary>MF_E_DRM_UNSUPPORTED.</summary>
    private const int DrmUnsupported = unchecked((int)0xC00D0BB8);

    public static async Task<VideoMetadata> ReadAsync(string path)
    {
        var name = Path.GetFileName(path);

        if (!File.Exists(path))
        {
            throw WallpaperException.FileMissing(name);
        }

        StorageFile file;
        try
        {
            file = await StorageFile.GetFileFromPathAsync(path);
        }
        catch (Exception error)
        {
            throw WallpaperException.ImportFailed(name, error.Message);
        }

        var properties = await file.Properties.GetVideoPropertiesAsync();
        var width = (int)properties.Width;
        var height = (int)properties.Height;
        var duration = properties.Duration.TotalSeconds;

        // Creating a clip is what actually proves the installed codecs can
        // decode this file; the shell properties above come from the container.
        MediaClip clip;
        try
        {
            clip = await MediaClip.CreateFromFileAsync(file);
        }
        catch (Exception error) when (error.HResult == DrmUnsupported)
        {
            throw WallpaperException.ProtectedContent(name);
        }
        catch (Exception error)
        {
            Log.Library.Error($"Clip creation failed for {name}: {error.Message}");
            throw WallpaperException.NotPlayable(name);
        }

        if (duration <= 0)
        {
            duration = clip.OriginalDuration.TotalSeconds;
        }

        if (width <= 0 || height <= 0)
        {
            throw WallpaperException.NoVideoTrack(name);
        }

        if (double.IsNaN(duration) || double.IsInfinity(duration) || duration <= 0)
        {
            throw WallpaperException.EmptyDuration(name);
        }

        return new VideoMetadata(duration, width, height);
    }
}
